### DISCUSSION TITLE
Help with checking one modified cystal plasticity program

### DISCUSSION THREAD

**Discussion by xchengood** [idaholab/moose#22665] [Q&A Modules: Solid mechanics] [answered]
Dear Moose experts or users,

I am trying to achieve the constitute model mentioned in section 4 of the literature [1] and section 2 of the literature [2] in Moose. I did some modifications in CrystalPlasticityStressUpdateBase.h, CrystalPlasticityStressUpdateBase.c, CrystalPlasticityKalidindiUpdate.h and CrystalPlasticityKalidindiUpdate.c. Specific changes are described below. All codes are here [https://github.com/xchengood/Nonschmid](url). Could anyone go to look through it and help me check that? The test file that I used is update_method_test.i. Thank you.

(1) In CrystalPlasticityStressUpdateBase.h [https://github.com/xchengood/Nonschmid/blob/main/CrystalPlasticityStressUpdateBase.h](url)

The resolved shear stress in equation 3 in Literature [2] is calculated by taking the non-Schmid effect into consideration. So there are some modifications in line 93 and line 234.

(2) In CrystalPlasticityStressUpdateBase.c [https://github.com/xchengood/Nonschmid/blob/main/CrystalPlasticityStressUpdateBase.C](url)

Modifications in line 107, line 126, line 130, line 371, line 413 - line 460, line 471, and line 483.
line 413 - 460 is to achieve equation (4) in literature [2].

(3) In CrystalPlasticityKalidindiUpdate.h https://github.com/xchengood/Nonschmid/blob/main/CrystalPlasticityKalidindiUpdate.h

Modifications in lines 101 - 106 express variables in equations (7) - (11) in literature [1].

(4) In CrystalPlasticityKalidindiUpdate.c [https://github.com/xchengood/Nonschmid/blob/main/CrystalPlasticityKalidindiUpdate.C](url)
Modifications are in lines 28 - 33, line3 54 - 59, line 117, line 185 - line 198.
line 117 is to achieve equation (10) in literature [1].
line 185 - line 198 is to achieve dslip_dtau.

[1]Yalcinkaya T, Brekelmans WA, Geers MG. BCC single crystal plasticity modeling and its experimental identification. Modelling and Simulation in Materials Science and Engineering. 2008 Oct 1;16(8):085007.
[2]Lim H, Hale LM, Zimmerman JA, Battaile CC, Weinberger CR. A multi-scale model of dislocation plasticity in α-Fe: Incorporating temperature, strain rate and non-Schmid effects. International Journal of Plasticity. 2015 Oct 1;73:100-18.

**Comment by hugary1995** [comment, marked as answer]
You can set up a single element test to probe the flow curve, and compare it with the curves reported in the papers. If there is any discrepancy, feel free to post your comparisons here and hopefully someone can help you diagnose that.
